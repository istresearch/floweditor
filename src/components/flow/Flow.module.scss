@import 'variables.module.scss';

.empty_flow {
  margin: 30px;
  width: 400px;

  h1 {
    font-family: 'Roboto', sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}