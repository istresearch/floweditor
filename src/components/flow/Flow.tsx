import { default as bindCallbacks } from 'auto-bind/react';
import Button from 'components/button/Button';
import { Canvas } from 'components/canvas/Canvas';
import { CanvasDraggableProps } from 'components/canvas/CanvasDraggable';
import Node from 'components/flow/node/Node';
import { getDraggedFrom } from 'components/helpers';
import NodeEditor from 'components/nodeeditor/NodeEditor';
import Simulator from 'components/simulator/Simulator';
import Sticky, { STICKY_BODY, STICKY_TITLE } from 'components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { FlowDefinition, FlowMetadata, FlowPosition } from 'flowTypes';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Plumber from 'services/Plumber';
import { DragSelection, DebugState } from 'store/editor';
import { RenderNode } from 'store/flowContext';
import { createEmptyNode, detectLoops, getOrderedNodes } from 'store/helpers';
import { NodeEditorSettings } from 'store/nodeEditor';
import AppState from 'store/state';
import {
  ConnectionEvent,
  DispatchWithState,
  mergeEditorState,
  MergeEditorState,
  NoParamsAC,
  onConnectionDrag,
  OnConnectionDrag,
  OnOpenNodeEditor,
  onOpenNodeEditor,
  onRemoveNodes,
  OnRemoveNodes,
  OnUpdateCanvasPositions,
  onUpdateCanvasPositions,
  resetNodeEditingState,
  UpdateConnection,
  updateConnection,
  updateSticky,
  UpdateSticky
} from 'store/thunks';
import {
  createUUID,
  isRealValue,
  NODE_PADDING,
  renderIf,
  snapToGrid,
  timeEnd,
  timeStart
} from 'utils';
import Debug from 'utils/debug';

import styles from './Flow.module.scss';
import { Trans } from 'react-i18next';
import { PopTabType } from 'config/interfaces';
import i18n from 'config/i18n';

declare global {
  interface Window {
    fe: any;
  }
}

export interface FlowStoreProps {
  ghostNode: RenderNode;
  debug: DebugState;
  translating: boolean;
  popped: string;
  dragActive: boolean;

  mergeEditorState: MergeEditorState;

  definition: FlowDefinition;
  nodes: { [uuid: string]: RenderNode };
  metadata: FlowMetadata;
  nodeEditorSettings: NodeEditorSettings;

  updateConnection: UpdateConnection;
  onOpenNodeEditor: OnOpenNodeEditor;
  onUpdateCanvasPositions: OnUpdateCanvasPositions;
  onRemoveNodes: OnRemoveNodes;
  resetNodeEditingState: NoParamsAC;
  onConnectionDrag: OnConnectionDrag;
  updateSticky: UpdateSticky;
}

export interface Translations {
  [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
  return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
  const left = Math.min(drag.startX, drag.currentX);
  const top = Math.min(drag.startY, drag.currentY);
  const width = Math.max(drag.startX, drag.currentX) - left;
  const height = Math.max(drag.startY, drag.currentY) - top;
  return {
    left,
    top,
    width,
    height
  };
};

export class Flow extends React.PureComponent<FlowStoreProps, {}> {
  private Plumber: Plumber;
  private nodeContainerUUID: string;

  // Refs
  private ghost: any;

  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: FlowStoreProps, context: ConfigProviderContext) {
    super(props, context);

    this.nodeContainerUUID = createUUID();

    this.Plumber = new Plumber();

    /* istanbul ignore next */
    if (context.config.debug) {
      window.fe = new Debug(props, this.props.debug);
    }

    bindCallbacks(this, {
      include: [/Ref$/, /^on/, /^is/, /^get/, /^handle/]
    });

    timeStart('Loaded Flow');
  }

  private ghostRef(ref: any): any {
    return (this.ghost = ref);
  }

  public componentDidMount(): void {
    this.Plumber.bind('connection', (event: ConnectionEvent) =>
      this.props.updateConnection(event.sourceId, event.targetId)
    );
    this.Plumber.bind('beforeDrag', (event: ConnectionEvent) => {
      this.beforeConnectionDrag(event);
    });

    this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
      this.props.onConnectionDrag(event, this.context.config.flowType);
    });

    this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
      this.onConnectorDrop(event)
    );
    this.Plumber.bind(
      'beforeStartDetach',
      (event: ConnectionEvent) => !this.props.translating && this.context.config.mutable
    );
    this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
    this.Plumber.bind('beforeDrop', (event: ConnectionEvent) => this.onBeforeConnectorDrop(event));
    this.Plumber.triggerLoaded(this.context.config.onLoad);

    timeEnd('Loaded Flow');
  }

  public componentWillUnmount(): void {
    this.Plumber.reset();
  }

  /**
   * Called right before a connector is dropped onto a new node
   */
  private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
    this.props.resetNodeEditingState();
    const fromNodeUUID = event.sourceId.split(':')[0];
    try {
      detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
    } catch {
      return false;
    }
    return true;
  }

  /**
   * Called the moment a connector is done dragging, whether it is dropped on an
   * existing node or on to empty space.
   */
  private onConnectorDrop(event: ConnectionEvent): boolean {
    const ghostNode = this.props.ghostNode;
    // Don't show the node editor if we a dragging back to where we were
    if (isRealValue(ghostNode) && !isDraggingBack(event)) {
      // Wire up the drag from to our ghost node
      this.Plumber.recalculate(ghostNode.node.uuid);

      const dragPoint = getDraggedFrom(ghostNode);

      this.Plumber.connect(dragPoint.nodeUUID + ':' + dragPoint.exitUUID, ghostNode.node.uuid);

      // Save our position for later
      const { left, top } = (this.ghost &&
        snapToGrid(this.ghost.ele.offsetLeft, this.ghost.ele.offsetTop)) || { left: 0, top: 0 };

      this.props.ghostNode.ui.position = { left, top };

      let originalAction = null;
      if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
        originalAction = ghostNode.node.actions[0];
      }

      // Bring up the node editor
      this.props.onOpenNodeEditor({
        originalNode: ghostNode,
        originalAction
      });
    }

    if (isDraggingBack(event)) {
      this.props.mergeEditorState({ ghostNode: null });
    }

    /* istanbul ignore next */
    document.removeEventListener('mousemove', (window as any).ghostListener);

    return true;
  }

  private beforeConnectionDrag(event: ConnectionEvent): boolean {
    if (event.source) {
      event.source.dispatchEvent(new Event('disconnect'));
    }
    return !this.props.translating;
  }

  private handleStickyCreation(props: CanvasDraggableProps) {
    const stickyMap = this.props.definition._ui.stickies || {};
    const uuid = props.uuid;
    return <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={props.selected} />;
  }

  private handleNodeCreation(props: CanvasDraggableProps) {
    const onlyNode = Object.keys(this.props.nodes).length === 1;
    return (
      <Node
        onlyNode={onlyNode}
        startingNode={props.idx === 0}
        selected={props.selected}
        key={props.uuid}
        data-spec={nodeSpecId}
        nodeUUID={props.uuid}
        plumberMakeTarget={this.Plumber.makeTarget}
        plumberRemove={this.Plumber.remove}
        plumberRecalculate={this.Plumber.recalculate}
        plumberMakeSource={this.Plumber.makeSource}
        plumberConnectExit={this.Plumber.connectExit}
        plumberUpdateClass={this.Plumber.updateClass}
      />
    );
  }

  private getNodes(): CanvasDraggableProps[] {
    return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
      return {
        uuid: renderNode.node.uuid,
        position: renderNode.ui.position,
        elementCreator: this.handleNodeCreation,
        config: renderNode,
        idx
      };
    });
  }

  private getStickies(): CanvasDraggableProps[] {
    const stickyMap = this.props.definition._ui.stickies || {};
    return Object.keys(stickyMap).map((uuid: string, idx: number) => {
      return {
        uuid,
        elementCreator: this.handleStickyCreation,
        position: stickyMap[uuid].position,
        idx
      };
    });
  }

  private getDragNode(): JSX.Element {
    return isRealValue(this.props.ghostNode) ? (
      <div
        data-spec={ghostNodeSpecId}
        key={this.props.ghostNode.node.uuid}
        style={{ position: 'absolute', display: 'block', visibility: 'hidden' }}
      >
        <Node
          onlyNode={false}
          selected={false}
          startingNode={false}
          ref={this.ghostRef}
          ghost={true}
          nodeUUID={this.props.ghostNode.node.uuid}
          plumberMakeTarget={this.Plumber.makeTarget}
          plumberRemove={this.Plumber.remove}
          plumberRecalculate={this.Plumber.recalculate}
          plumberMakeSource={this.Plumber.makeSource}
          plumberConnectExit={this.Plumber.connectExit}
          plumberUpdateClass={this.Plumber.updateClass}
        />
      </div>
    ) : null;
  }

  private getSimulator(): JSX.Element {
    return renderIf(this.context.config.endpoints && this.context.config.endpoints.simulateStart)(
      <Simulator
        key="simulator"
        popped={this.props.popped}
        mergeEditorState={this.props.mergeEditorState}
        onToggled={(visible: boolean, tab: PopTabType) => {
          this.props.mergeEditorState({
            popped: visible ? tab : null
          });
        }}
      />
    );
  }

  private getNodeEditor(): JSX.Element {
    return renderIf(this.props.nodeEditorSettings !== null)(
      <NodeEditor
        key="node-editor"
        helpArticles={this.context.config.help}
        plumberConnectExit={this.Plumber.connectExit}
      />
    );
  }

  // TODO: this should be a callback from the canvas
  private handleDoubleClick(position: FlowPosition): void {
    const { left, top } = position;
    this.props.updateSticky(createUUID(), {
      position: snapToGrid(left - 90 + NODE_PADDING, top - 40),
      title: STICKY_TITLE,
      body: STICKY_BODY
    });
  }

  private getEmptyFlow(): JSX.Element {
    return (
      <div key="create_node" className={styles.empty_flow}>
        <Trans i18nKey="empty_flow_message">
          <h1>Let's get started</h1>
          <div>
            We recommend starting your flow by sending a message. This message will be sent to
            anybody right after they join the flow. This is your chance to send a single message or
            ask them a question.
          </div>
        </Trans>

        <Button
          name={i18n.t('buttons.create_message', 'Create Message')}
          onClick={() => {
            const emptyNode = createEmptyNode(null, null, 1, this.context.config.flowType);
            this.props.onOpenNodeEditor({
              originalNode: emptyNode,
              originalAction: emptyNode.node.actions[0]
            });
          }}
        />
      </div>
    );
  }

  /* 
  public componentDidUpdate(prevProps: FlowStoreProps): void {
    traceUpdate(this, prevProps);
  }
  */

  public handleDragging(uuids: string[]): void {
    uuids.forEach((uuid: string) => {
      try {
        const ele = document.getElementById(uuid);
        const exits = ele.querySelectorAll('.jtk-connected');
        this.Plumber.revalidate([ele, ...exits]);
      } catch (error) {}
    });
  }

  public handleCanvasLoaded(): void {
    this.Plumber.setContainer('canvas');
  }

  public render(): JSX.Element {
    const nodes = this.getNodes();

    const draggables = this.getStickies().concat(nodes);

    return (
      <div>
        {nodes.length === 0 ? this.getEmptyFlow() : <>{this.getSimulator()}</>}
        {this.getNodeEditor()}

        <Canvas
          mutable={this.context.config.mutable}
          draggingNew={!!this.props.ghostNode && !this.props.nodeEditorSettings}
          newDragElement={this.getDragNode()}
          onDragging={this.handleDragging}
          uuid={this.nodeContainerUUID}
          dragActive={this.props.dragActive}
          mergeEditorState={this.props.mergeEditorState}
          onRemoveNodes={this.props.onRemoveNodes}
          draggables={draggables}
          onDoubleClick={this.handleDoubleClick}
          onUpdatePositions={this.props.onUpdateCanvasPositions}
          onLoaded={this.handleCanvasLoaded}
        ></Canvas>
        <div id="activity_recent_contacts"></div>
      </div>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = ({
  flowContext: { definition, metadata, nodes },
  editorState: { ghostNode, debug, translating, popped, dragActive },
  // tslint:disable-next-line: no-shadowed-variable
  nodeEditor: { settings }
}: AppState) => {
  return {
    nodeEditorSettings: settings,
    definition,
    nodes,
    metadata,
    ghostNode,
    debug,
    translating,
    popped,
    dragActive
  };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators(
    {
      mergeEditorState,
      resetNodeEditingState,
      onConnectionDrag,
      onOpenNodeEditor,
      onUpdateCanvasPositions,
      onRemoveNodes,
      updateConnection,
      updateSticky
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Flow);
