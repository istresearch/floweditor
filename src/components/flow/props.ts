import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties, FlowIssue } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';

export interface IssueProps {
  helpArticles: { [key: string]: string };
  issues: FlowIssue[];
}

export interface ActionFormProps extends IssueProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps extends IssueProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps extends IssueProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): void;
  onClose(canceled: boolean): void;
  helpArticles: { [key: string]: string };
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};

export const STATUS_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Status),
  id: ContactProperties.Status,
  type: AssetType.ContactProperty
};
