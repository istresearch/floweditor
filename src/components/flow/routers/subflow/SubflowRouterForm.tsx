import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from 'components/flow/routers/subflow/helpers';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import * as React from 'react';
import { Asset } from 'store/flowContext';
import { FormEntry, FormState, mergeForm, StringEntry } from 'store/nodeEditor';
import { shouldRequireIf, validate } from 'store/validators';
import i18n from 'config/i18n';
import { fetchAsset, getFlowType } from 'external';
import styles from './SubflowRouterForm.module.scss';
import { Trans } from 'react-i18next';
import TextInputElement from 'components/form/textinput/TextInputElement';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';

// TODO: Remove use of Function
export interface SubflowRouterFormState extends FormState {
  flow: FormEntry;
  params: { [name: string]: StringEntry };
}

export default class SubflowRouterForm extends React.PureComponent<
  RouterFormProps,
  SubflowRouterFormState
> {
  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  public componentDidMount() {
    // we need to resolve our flow for it's parent refs
    // todo: just fetch this is a plan flow result without the asset translation
    if (this.state.flow.value) {
      fetchAsset(this.props.assetStore.flows, this.state.flow.value.uuid).then((flow: Asset) => {
        if (flow) {
          this.handleFlowChanged([
            { name: flow.name, uuid: flow.id, parent_refs: flow.content.parent_refs }
          ]);
        }
      });
    }
  }

  public handleFlowChanged(flows: any[], submitting = false): boolean {
    const flow = flows[0];

    const updates: Partial<SubflowRouterFormState> = {
      flow: validate(i18n.t('forms.flow', 'Flow'), flow, [shouldRequireIf(submitting)])
    };

    const params: { [key: string]: StringEntry } = {};
    // ensure our parameters are initialized
    if (flow && flow.parent_refs) {
      for (const key of flow.parent_refs) {
        if (this.state.params[key]) {
          params[key] = { ...this.state.params[key] };
        } else {
          params[key] = { value: '' };
        }
      }
      updates.params = params;
    }

    const updated = mergeForm(this.state, updates);
    this.setState(updated);

    return updated.valid;
  }

  private handleSave(): void {
    // validate our flow in case they haven't interacted
    this.handleFlowChanged([this.state.flow.value], true);

    const hasFieldErrors = Object.keys(this.state.params).find((key: string) =>
      hasErrors(this.state.params[key])
    );

    if (this.state.valid && !hasFieldErrors) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private handleShouldExclude(flow: any): boolean {
    // only show flows that match our flow type
    return getFlowType(flow) !== this.context.config.flowType;
  }

  private handleParamChanged(text: string, name: string) {
    const params = { ...this.state.params };
    params[name] = { value: text };
    this.setState({ params });
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const tabs: Tab[] = [];
    const flow = this.state.flow.value;

    const hasFieldErrors = !!Object.keys(this.state.params).find((key: string) =>
      hasErrors(this.state.params[key])
    );

    if (flow && flow.parent_refs && flow.parent_refs.length > 0) {
      tabs.push({
        name: i18n.t('forms.enter_flow_parameters_tab', 'Parameters'),
        body: (
          <div>
            <p className={styles.info}>
              <Trans
                i18nKey="forms.enter_flow_parameters_summary"
                values={{
                  flow: this.state.flow.value.name,
                  url: this.context.config.endpoints.editor + '/' + this.state.flow.value.id
                }}
              >
                <a
                  target="_"
                  href={this.context.config.endpoints.editor + '/' + this.state.flow.value.id}
                >
                  [[flow]]
                </a>{' '}
                expects the following parameters to be set by this flow. These can be set using a{' '}
                <span>Save Flow Result</span> action or directly below.
              </Trans>
            </p>
            <table className={styles.params}>
              <tbody>
                {flow.parent_refs.map((name: string) => {
                  return (
                    <tr key={'param_' + name} className={styles.param}>
                      <td className={styles.param_name}>{name}</td>
                      <td className={styles.param_input}>
                        <TextInputElement
                          name={name}
                          showLabel={false}
                          placeholder={name}
                          onChange={(updatedText: string) => {
                            this.handleParamChanged(updatedText, name);
                          }}
                          entry={this.state.params[name]}
                          autocomplete={true}
                        />
                      </td>
                    </tr>
                  );
                })}
              </tbody>
            </table>
          </div>
        ),
        hasErrors: hasFieldErrors,
        checked: !!Object.keys(this.state.params).find(
          (key: string) => this.state.params[key] && this.state.params[key].value.trim().length > 0
        )
      });
    }

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <AssetSelector
          name={i18n.t('forms.flow', 'Flow')}
          placeholder={i18n.t('forms.select_flow', 'Select the flow to start')}
          assets={this.props.assetStore.flows}
          entry={this.state.flow}
          searchable={true}
          shouldExclude={this.handleShouldExclude}
          onChange={this.handleFlowChanged}
        />
        {renderIssues(this.props)}
      </Dialog>
    );
  }
}
