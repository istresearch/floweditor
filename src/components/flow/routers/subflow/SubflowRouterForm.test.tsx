import { RouterFormProps } from 'components/flow/props';
import { Types } from 'config/interfaces';
import { createSubflowNode, createStartFlowAction } from 'testUtils/assetCreators';
import { getTypeConfig } from 'config';
import {
  render,
  mock,
  fireEvent,
  wait,
  fireChangeText,
  getUpdatedNode,
  getByTestId
} from 'test/utils';
import * as React from 'react';
import * as utils from 'utils';
import { RenderNode, AssetType } from 'store/flowContext';
import SubflowRouterForm from './SubflowRouterForm';
import * as external from 'external';

mock(utils, 'createUUID', utils.seededUUIDs());

const getRouterFormProps = (type: Types, originalNode: RenderNode): RouterFormProps => {
  return {
    nodeSettings: {
      originalNode
    },
    typeConfig: getTypeConfig(type),
    assetStore: { flows: { items: {}, type: AssetType.Flow } },
    updateRouter: jest.fn(),
    onTypeChange: jest.fn(),
    onClose: jest.fn()
  };
};

const subflowProps = getRouterFormProps(
  Types.split_by_subflow,
  createSubflowNode(createStartFlowAction())
);

mock(
  external,
  'fetchAsset',
  utils.fetchAsset({
    id: 'my-subflow',
    name: 'My Subflow Flow',
    type: AssetType.Flow,
    content: {
      parent_refs: ['max', 'min']
    }
  })
);

describe(SubflowRouterForm.name, () => {
  it('should render', () => {
    const { baseElement } = render(<SubflowRouterForm {...subflowProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should init parameter tab', async () => {
    const { baseElement } = render(<SubflowRouterForm {...subflowProps} />);
    // let our mount event to render our tabs
    await wait();
    expect(baseElement).toMatchSnapshot();
  });

  it('should create result actions for parameters', async () => {
    const { getByText, queryAllByTestId, getByTestId } = render(
      <SubflowRouterForm {...subflowProps} />
    );
    await wait();

    // open the parameter tab
    fireEvent.click(getByText('Parameters'));

    const min = getByTestId('min');
    const max = getByTestId('max');

    // enter some values
    fireChangeText(min, '1');
    fireChangeText(max, '100');

    fireEvent.click(getByText('Ok'));
    let actions = getUpdatedNode(subflowProps).node.actions;

    // should have a run result action for each parameter, plus an enter flow action
    expect(actions.length).toEqual(3);
    expect(actions[0].type).toEqual(Types.set_run_result);
    expect(actions[1].type).toEqual(Types.set_run_result);
    expect(actions[2].type).toEqual(Types.enter_flow);
    expect(subflowProps.updateRouter).toMatchCallSnapshot();

    // remove a parameter
    fireChangeText(min, '');
    fireEvent.click(getByText('Ok'));

    // now only two actions
    actions = getUpdatedNode(subflowProps).node.actions;
    expect(actions.length).toEqual(2);
    expect(actions[0].type).toEqual(Types.set_run_result);
    expect(actions[1].type).toEqual(Types.enter_flow);
  });
});
