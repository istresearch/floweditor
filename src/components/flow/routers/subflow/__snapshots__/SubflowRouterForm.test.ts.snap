// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubflowRouterForm render should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="enter_flow"
  tabs={Array []}
  title="Enter a Flow"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_subflow",
        ],
        "component": [Function],
        "description": "Enter another flow",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Enter a Flow",
        "type": "enter_flow",
      }
    }
    onChange={[MockFunction]}
  />
  <AssetSelector
    assets={
      Object {
        "items": Object {},
        "type": "flow",
      }
    }
    entry={
      Object {
        "value": Object {
          "name": "Colors",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      }
    }
    name="Flow"
    onChange={[Function]}
    placeholder="Select the flow to start"
    searchable={true}
    shouldExclude={[Function]}
  />
</Dialog>
`;

exports[`SubflowRouterForm updates converts from other node types 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Color",
            "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
          },
          "type": "enter_flow",
          "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "completed",
            ],
            "category_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "type": "has_only_text",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
          Object {
            "arguments": Array [
              "expired",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            "name": "Complete",
            "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          },
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Expired",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
        ],
        "default_category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        "operand": "@child.run.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates creates its own action uuid if necessary 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Color",
            "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
          },
          "type": "enter_flow",
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
        },
        Object {
          "destination_uuid": null,
          "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "completed",
            ],
            "category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
            "type": "has_only_text",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
          Object {
            "arguments": Array [
              "expired",
            ],
            "category_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "type": "has_only_text",
            "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            "name": "Complete",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          Object {
            "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "name": "Expired",
            "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
          },
        ],
        "default_category_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        "operand": "@child.run.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates should update and save 1`] = `
Object {
  "flow": Object {
    "validationFailures": Array [],
    "value": Object {
      "name": "Favorite Color",
      "parent_refs": Array [
        "colors",
      ],
      "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
    },
  },
  "params": Object {
    "colors": Object {
      "value": "",
    },
  },
  "valid": true,
}
`;

exports[`SubflowRouterForm updates should update and save 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Color",
            "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
          },
          "type": "enter_flow",
          "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
        Object {
          "destination_uuid": null,
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "C",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_run_status",
            "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
          },
          Object {
            "arguments": Array [
              "E",
            ],
            "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "type": "has_run_status",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Complete",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
          Object {
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "name": "Expired",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        "operand": "@child.run.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;
